### PR TITLE
No mapper service and index caches for replicated closed indices

### PR DESCRIPTION
**Pull request by ywelsch**
Replicated closed indices can't be indexed into or searched, and therefore don't need a shard with full indexing and search capabilities allocated. We can save on a lot of heap memory for those indices by not allocating a mapper service and caching infrastructure (which preallocates a constant amount per instance). Before this change, a 1GB ES instance could host 250 replicated closed metricbeat indices (each index with one shard). After this change, the same instance can host 7300 replicated closed metricbeat instances (not that this would be a recommended configuration). Most of the remaining memory is in the cluster state and the `IndexSettings` object.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/IndexService.java [hunk 1]
```diff
@@ -548,7 +565,10 @@ public ScriptService getScriptService() {
 
     @Override
     public boolean updateMapping(final IndexMetaData currentIndexMetaData, final IndexMetaData newIndexMetaData) throws IOException {
-        return mapperService().updateMapping(currentIndexMetaData, newIndexMetaData);
+        if (mapperService == null) {
+            return false;
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/IndexService.java]: Can we assert the state of `currentIndexMetaData` and `newIndexMetaData`?

**Review by dnhatn — APPROVED** [review]
Great change!

File: server/src/main/java/org/elasticsearch/index/IndexService.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/IndexService.java]: What would you assert about those?

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/IndexService.java]: I thought about asserting the index is closed but MapperService is set once so just ignore this.

**Pull request merged**